### ISSUE TITLE
Deriving Show fails to discover local fixity declaration

### ISSUE THREAD

**Issue by RyanGlScott** [goldfirere/singletons#192]
This file should typecheck, but doesn't:

```haskell
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE InstanceSigs #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE TemplateHaskell #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeInType #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE UndecidableInstances #-}
module Bug where

import Data.Singletons.Prelude
import Data.Singletons.TH
import Data.Type.Equality

$(singletons [d|

  infixl 6 :*:
  data T = () :*: () deriving Show

  |])

test1 :: "() :*: ()" :~: ShowsPrec 6 ('() :*: '()) ""
test1 = Refl

test2 :: "(() :*: ())" :~: ShowsPrec 7 ('() :*: '()) ""
test2 = Refl
```

```
GHCi, version 8.2.0.20170522: http://www.haskell.org/ghc/  :? for help
Loaded GHCi configuration from /home/rgscott/.ghci
[1 of 1] Compiling Bug              ( Bug.hs, interpreted )

Bug.hs:26:9: error:
    • Couldn't match type ‘"(() :*: ())"’ with ‘"() :*: ()"’
      Expected type: "(() :*: ())" :~: ShowsPrec 7 ('() ':*: '()) ""
        Actual type: "(() :*: ())" :~: "(() :*: ())"
    • In the expression: Refl
      In an equation for ‘test2’: test2 = Refl
   |
26 | test2 = Refl
   |         ^^^^
```

In plain terms, since the fixity of `(:*:)` is 6, then showing it with a precedence of 6 should not surround the output with parentheses (which is working), but showing it with a precedence of 7 _should_ surround the output with parentheses (which is not working).

The culprit is that `deriving Show` is failing to discover the local fixity declaration for `(:*:)`. This can be seen if you inspect the `-ddump-splices` output:

```haskell
type family ShowsPrec_6 (p :: Nat) (x :: T) (s :: Symbol) :: Symbol where
  ShowsPrec_6 p ((:*:) argL argR) x = Apply (Apply (Apply ShowParenSym0 (Apply (Apply (:>$) p) (FromInteger 9))) (Apply (Apply (:.$) (Apply (Apply ShowsPrecSym0 (FromInteger 10)) argL)) (Apply (Apply (:.$) (Apply ShowStringSym0 " :*: ")) (Apply (Apply ShowsPrecSym0 (FromInteger 10)) argR)))) x
```

Notice the `(Apply (Apply (:>$) p) (FromInteger 9))` part. That's totally bogus, since we should be using 6 there! This is doubly perplexing, since I'm using `reifyFixityWithLocals` in [the implementation of `deriving Show`](https://github.com/goldfirere/singletons/blob/75dae57bbe06b21b3adbb82158c4767ebd695d14/src/Data/Singletons/Deriving/Show.hs#L57):

```haskell
    -- Infix constructors have special Show treatment.
    go (DNormalC True [_, _]) = do
      argL <- newUniqueName "argL"
      argR <- newUniqueName "argR"
      fi <- fromMaybe defaultFixity <$> reifyFixityWithLocals con_name
      ...
```

For some reason, `reifyFixityWithLocals` must be failing, causing it to fall back to `defaultFixity`. But the reason why eludes me.

**Comment by RyanGlScott** [comment]
After thinking over this some more, I believe solving this problem would be tantamount to doing dependency analysis in Template Haskell. The example I posted above appears simple - after all, the fixity declaration `infixl 6 :*:` occurs before the corresponding data declaration, so you could "just" remember all the fixity declarations when processing `Dec`s. But this gets far more complicated when you consider that the fixity declaration could occur _after_ the data declaration...

I certainly don't want to implement dependency analysis (and I doubt you do either), so I think I would be content with being able to use standalone-derived `Show` instances in a separate Template Haskell splice (which #194 implements).

**Comment by goldfirere** [comment]
Why the need for dependency analysis? The `WithLocals` stuff looks at all the declarations in the splice, regardless of order. (See the call to `withLocalDeclarations` in `singTopLevelDecs`. So this looks like a th-desugar bug to me. Possibly in `nameMatches`.

**Comment by RyanGlScott** [comment]
Oh, _duh_. I just wasn't calling `mkShowInstances` within the scope of a `withLocalDeclarations` call! Luckily, that's easily fixed - see #196.
